bluemira/power_cycle/base.py

```diff
@@ -0,0 +1,468 @@
+# POWER CYCLE ABSTRACT BASE CLASS
+# ######################################################################
+
+
+class PowerCycleABC(abc.ABC):
+    """
+    Abstract base class for classes in the Power Cycle module.
+
+    Parameters
+    ----------
+    name: `str`
+        Description of the instance.
+    """
+
+    # ------------------------------------------------------------------
+    # CLASS ATTRIBUTES
+    # ------------------------------------------------------------------
+
+    # Plot defaults (arguments for `matplotlib.pyplot.plot`)
+    _plot_kwargs = {
+        "c": "k",  # Line color
+        "lw": 2,  # Line width
+        "ls": "-",  # Line style
+    }
+    _scatter_kwargs = {
+        "c": "k",  # Marker color
+        "s": 100,  # Marker size
+        "marker": "x",  # Marker style
+    }
+
+    # Plot text settings (for `matplotlib.pyplot.text`)
+    _text_angle = 45  # rotation angle
+    _text_index = -1  # index of (time,data) point used for location
+
+    # ------------------------------------------------------------------
+    # CONSTRUCTOR
+    # ------------------------------------------------------------------
+
+    def __init__(self, name: str):
+
+        # Store name
+        self.name = self._validate_name(name)
+
+        # Dynamically add validation classmethod
+        # setattr(a, 'func', classmethod(func))
+
+    @classmethod
+    def _validate_name(cls, name):
+        """
+        Validate a name for class instance creation to be a string.
+        """
+        class_name = cls.__name__
+        if not isinstance(name, (str)):
+            raise TypeError(
+                f"""
+                The 'name' used to create an instance of the
+                {class_name} class must be a `str`.
+                """
+            )
+
+        return name
+
+    @classmethod
+    def _issue_error(cls, label):
+        """
+        Issue error associated with...
+        """
+
+        # Retrieve errors of that class
+        error_dict = cls._errors
+
+        # Validate error label
+        all_labels = error_dict.keys()
+        if label in all_labels:
+
+            # Retrieve particular error
+            the_error = error_dict[label]
+
+            # Retrieve error attributes
+            error_type = the_error.err_type
+            error_msg = the_error.err_msg
+
+            # Build raising function
+            raise_function = error_type + "Error"
+            raise_function = f"""
+                raise {raise_function}('''{error_msg}''')
+                """
+
+            exec(raise_function)  # Issue error
+        else:
+
+            # Child class name
+            class_name = cls.__name__
+
+            # Issue error
+            raise ValueError(
+                f"""
+                Unknown error label for error in class {class_name}.
+                """
+            )
+
+    @staticmethod
+    def _validate_list(input) -> list:
```

Comment:
For the few places this could be needed just add it to the functions that do. You shouldnt try and type protect too vigorously.

Reply:
Got it, will remove it in the new commit.